jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebInfConfiguration.java

```diff
@@ -1003,11 +1037,53 @@ protected Resource findWebInfClassesDir(WebAppContext context)
         StringTokenizer tokenizer = new StringTokenizer(context.getExtraClasspath(), ",;");
         while (tokenizer.hasMoreTokens())
         {
-            Resource resource = context.newResource(tokenizer.nextToken().trim());
-            if (resource.exists() && resource.isDirectory())
-                dirResources.add(resource);
+            String token = tokenizer.nextToken().trim();
+            if (isGlobReference(token))
+            {
```

Comment:
If the entry has a trailing *, then it is a directory of jar files and we can ignore it, we should not look deeply into its subdirs.

Reply:
We only look at the level of the `/*`.
So that means if we have say ...

``` shell
$ ls -1 -F /opt/ext/
corp.jar
foo.jar
resources/
```

And a reference of `/opt/ext/*` then that's 3 URL entries in the WebAppClassLoader.

1. file:///opt/ext/corp.jar
2. file:///opt/ext/foo.jar
3. file:///opt/ext/resources/

Reply:
This is not the WebAppClassLoader, this is in WebInfConfiguration.findExtraClasspathDirs where we are only looking for entries in the ;, separated list that are themselves directories; any entry that is suffixed by /* is handled by WebInfConfiguration.findExtraClasspathJars. We don't want to do a deep add of a directory inside of WebInfConfiguration.findExtraClasspathDirs.

Reply:
This doesn't do deep directory, it's only at the level of the glob.

Glob means anything relevant in the point of the tree of the glob, not deeper.

"Relevant" is anything that has meaning to the Classloader, which is zip, jar, and directory (all other entries that the glob produces are ignored).

Reply:
So, we have a misaligned behavior here.

`WebInfConfiguration` and `WebAppClassLoader` in existing jetty-9.4.x behaves differently in regards to `/path/to/*` entries.

We either use the `WebAppClassLoader` behavior for glob that adds directories.
Or we use the `WebInfConfiguration` behavior that glob doesn't add directories.

Looking at how java itself works I think we should use the existing `jetty-9.4.x` behavior that `WebInfConfiguration` uses and not add directories.

I'll explain with a code example ...

``` java
package cloader;

import java.io.File;

public class ShowClassLoader
{
    public static void main(String[] args)
    {
        String javaClasspath = System.getProperty("java.class.path");
        System.out.printf("java.class.path=%s%n", javaClasspath);
        String[] parts = javaClasspath.split(File.pathSeparator);
        System.out.printf("  entries.length=%d%n", parts.length);
        for (int i = 0; i < parts.length; i++)
        {
            System.out.printf("  entry[%d]=%s%n", i, parts[i]);
        }
    }
}
```

Run on the command line ...

``` shell
$ ls -1 -F target/lib
corp.jar
foo.jar
resources/

$ java -cp 'target/lib/*:target/classes' cloader.ShowClassLoader
java.class.path=target/lib/corp.jar:target/lib/foo.jar:target/classes
  entries.length=3
  entry[0]=target/lib/corp.jar
  entry[1]=target/lib/foo.jar
  entry[2]=target/classes
```

This behavior on OpenJDK 11.0.7 is what `WebInfConfiguration` mimics, but not what `WebAppClassLoader` does.

The OpenJDK 11.0.7 on Windows classpath behaves the same way.

And it's documented like this at https://docs.oracle.com/javase/8/docs/technotes/tools/windows/classpath.html#A1100762

Reply:
Just as a followup, the glob behaviors on Linux from the above example seem to be mostly done via the shell.

Windows JVM seems to have specific code to allow globs, so we are essentially attempting to mimic glob shell behavior on linux and glob classpath behavior on windows.

Reply:
This has been revered in WebInfConfiguration and WebAppClassloader has been updated to reflect the truth in both JVM behavior and WebInfConfiguration behavior (along with tests).

Reply:
So the point I was making was that in WebInfConfiguration.findExtraClasspathJars we should not add the directory with the trailing * to the list of resources, only the jars and zips that are contained at the top level inside of that directory. If you've changed the code so that that directory is _not_ added, then we're good.

Reply:
Current implementation in this PR does not add directories for glob usage in either WebInfConfiguration (like before) nor in WebAppClassLoader (this is a change in behavior)